File: TinyScript/java/src/main/java/parser/util/PeekTokenIterator.java
package parser.util;

import common.PeekIterator;
import lexer.Token;
import lexer.TokenType;

import java.util.stream.Stream;

public class PeekTokenIterator extends PeekIterator<Token> {

    public PeekTokenIterator(Stream<Token> stream) {
        super(stream);
    }

    public Token nextMatch(String value) throws ParseException {
        var token = this.next();
        if(!token.getValue().equals(value)) {
            throw new ParseException(token);
        }
        return token;
    }

    public Token nextMatch(TokenType type) throws ParseException {
        var token = this.next();
        if(!token.getType().equals(type)) {
            throw new ParseException(token);
        }
        return token;
    }
}

File: TinyScript/js/src/parser/util/PeekTokenIterator.js
const PeekIterator = require('../../common/PeekIterator')
const ParseException = require('./ParseException')

class PeekTokenIterator extends PeekIterator{
    constructor(it) {
        super(it)
    }

    nextMatch(value) {
        const token = this.next()
        if(token.getValue() !== value) {
            throw ParseException.fromToken(token)
        }
        return token
    }

    nextMatch1(type) {
        const token = this.next()
        if(token.getType() !== type) {
            throw ParseException.fromToken(token)
        }
        return token
    }
}

module.exports = PeekTokenIterator